8348427: DeferredLintHandler API should use JCTree instead of DiagnosticPosition

archiecobbs opened the pull request:
The purpose of `DeferredLintHandler` is to allow `@SuppressWarnings` to be applied to warnings that are generated before `@SuppressWarnings` annotations themselves have been processed. The way this currently works is that warning callbacks are kept in a `HashMap` keyed by the innermost containing module, package, class, method, or variable declarations (in the form of a `JCModuleDecl`, `JCPackageDecl`, `JCClassDecl`, `JCMethodDecl`, or `JCVariableDecl`). Later, when the compiler executes the attribution phase and the lint categories suppressed at each declaration are known, the corresponding warning callbacks are provided with an appropriately configured `Lint` instance and "flushed".

However, the `DeferredLintHandler` API uses `DiagnosticPosition` instead of `JCTree` for registering and flushing deferred warnings. This opens the door for bugs where warnings are registered to an object which is not a declaration, and therefore ignored.

In fact, this occurs once in the code ([here](https://github.com/openjdk/jdk/blob/48ece0721489c1b357aaa81e89fe59f486079d15/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Annotate.java#L679)) where a `JCExpression` is being passed, although in this case the bug appears to be harmless (because annotation values can't contain any type of declaration).

The API should be tighted up, and furthermore an assertion should be added to verify that the JCTree being passed is actually a declaration supporting `@SuppressWarnings`.

In addition, there is a design flaw in the API: it's not possible to obtain the current immediate mode `Lint` object, so if an immediate mode `Lint` object is pushed/popped more than once, the second `Lint` object will overwrite the first.

To fix this, the API should be adjusted so the stack of current declarations and/or immediate mode `Lint` objects is managed by the `DeferredLintHandler` itself, by providing a `push()` and `pop()` methods in the API.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8348427](https://bugs.openjdk.org/browse/JDK-8348427): DeferredLintHandler API should use JCTree instead of DiagnosticPosition (**Enhancement** - P4)


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23281/head:pull/23281` \
`$ git checkout pull/23281`

Update a local copy of the PR: \
`$ git checkout pull/23281` \
`$ git pull https://git.openjdk.org/jdk.git pull/23281/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23281`

View PR using the GUI difftool: \
`$ git pr show -t 23281`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23281.diff">https://git.openjdk.org/jdk/pull/23281.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23281#issuecomment-2611084075)
</details>

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/code/DeferredLintHandler.java:
```diff
@@ -52,99 +71,107 @@ public static DeferredLintHandler instance(Context context) {
-            loggers.append(logger);
-        }
+    public void push(JCTree decl) {
+        Assert.check(decl.getTag() == Tag.MODULEDEF
```
Nice!

mcimadamore commented:
Question: looking more broadly at the deferred lint handler API, I see that in most cases (all but one) `report` is called with a lambda that ignores its `lint` object. This seems odd - consider this:

```
private void warnOnExplicitStrictfp(DiagnosticPosition pos) {
        DiagnosticPosition prevLintPos = deferredLintHandler.setPos(pos);
        try {
            deferredLintHandler.report(_ -> lint.logIfEnabled(log, pos, LintWarnings.Strictfp)); // <---------------
        } finally {
            deferredLintHandler.setPos(prevLintPos);
        }
    }
```

Is the above code correct? E.g. should the call on `lint.logIfEnabled` occur on the `lint` that is the parameter to the lambda expression?

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/code/DeferredLintHandler.java:
```diff
@@ -52,99 +71,107 @@ public static DeferredLintHandler instance(Context context) {
-            }
-            loggers.append(logger);
-        }
+    public void push(JCTree decl) {
```
I wonder if `enterDecl`/`leaveDecl` would be more evocative than `push`/`pop`. (but don't have any suggestion for the weird immediate mode).

mcimadamore commented:
(My doubts re DeferredLintHandler are whether it's a good abstraction or not. It seems to provide, on paper, a good service: allow for lint warnings to be correctly reported in earlier phases of the compiler. However, I note that:
* this class seems to be serving too many masters -- the "immediate" mode seems to be a sign of that
* using this class correctly is rather finicky, and depends on which part of javac you are on. For instance, Flow doesn't use it, because it is after Attr... in other words, reporting of lint warnings is not centralized - and each components does what it needs to get the job done -- understandable, but leads to messy code (this also leads to Check::setLint)
* the synergy between Lint and DeferredLintHandler goes quite deep. For instance, flush has to be called in a place where `Lint` has been augmented with the contents of the relevant `SuppressWarnings`. All these dependencies make it quite hard to verify that the code does what it needs to do.

My general feeling here is that we'd be better off with a separate compilation step that runs after all the various front-end steps (e.g. after Flow). E.g. up to Flow, we keep accumulating lint warnings grouped by declaration. Then, in a single pass we take care of updating Lint with the correct `@SuppressWarning` found in a given declaration, and flush all the pending lint warnings for all that declaration. This would also mean that Attr and Check would be hypothetically be freed from Lint duties. Maybe this is too extreme, and we can't get there in one step - but I'd like to know what you think about the general direction (since we have already added more compilation steps in this area).

archiecobbs commented:
> Is the above code correct? E.g. should the call on `lint.logIfEnabled` occur on the `lint` that is the parameter to the lambda expression?

I wondered the same thing, and it seems plainly counter to the whole point of warning deferral. I haven't tried to figure it out yet - I assume it must work because when the lambda executes, the correct `Lint` object happens to still be in place.

But you got me curious... my theory is easy to test with this patch:
```diff
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java
index fe40bf0dad1..3a23c5df6d6 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java
@@ -244,7 +244,7 @@ MethodSymbol setMethod(MethodSymbol newMethod) {
      *  @param pos        Position to be used for error reporting.
      *  @param sym        The deprecated symbol.
      */
-    void warnDeprecated(DiagnosticPosition pos, Symbol sym) {
+    void warnDeprecated(Lint lint, DiagnosticPosition pos, Symbol sym) {
         if (sym.isDeprecatedForRemoval()) {
             if (!lint.isSuppressed(LintCategory.REMOVAL)) {
                 if (sym.kind == MDL) {
@@ -284,7 +284,7 @@ public void warnDeclaredUsingPreview(DiagnosticPosition pos, Symbol sym) {
      *  @param pos        Position to be used for error reporting.
      *  @param msg        A Warning describing the problem.
      */
-    public void warnRestrictedAPI(DiagnosticPosition pos, Symbol sym) {
+    public void warnRestrictedAPI(Lint lint, DiagnosticPosition pos, Symbol sym) {
         lint.logIfEnabled(log, pos, LintWarnings.RestrictedMethod(sym.enclClass(), sym));
     }
 
@@ -648,8 +648,8 @@ public void checkRedundantCast(Env<AttrContext> env, final JCTypeCast tree) {
                 && types.isSameType(tree.expr.type, tree.clazz.type)
                 && !(ignoreAnnotatedCasts && TreeInfo.containsTypeAnnotation(tree.clazz))
                 && !is292targetTypeCast(tree)) {
-            deferredLintHandler.report(_l -> {
-                lint.logIfEnabled(log, tree.pos(), LintWarnings.RedundantCast(tree.clazz.type));
+            deferredLintHandler.report(lint2 -> {
+                lint2.logIfEnabled(log, tree.pos(), LintWarnings.RedundantCast(tree.clazz.type));
             });
         }
     }
@@ -1324,7 +1324,7 @@ && checkDisjoint(pos, flags,
     private void warnOnExplicitStrictfp(JCTree tree) {
         deferredLintHandler.push(tree);
         try {
-            deferredLintHandler.report(_ -> lint.logIfEnabled(log, tree.pos(), LintWarnings.Strictfp));
+            deferredLintHandler.report(lint2 -> lint2.logIfEnabled(log, tree.pos(), LintWarnings.Strictfp));
         } finally {
             deferredLintHandler.pop();
         }
@@ -3785,7 +3785,7 @@ void checkDeprecated(Supplier<DiagnosticPosition> pos, final Symbol other, final
                 || s.isDeprecated() && !other.isDeprecated())
                 && (s.outermostClass() != other.outermostClass() || s.outermostClass() == null)
                 && s.kind != Kind.PCK) {
-            deferredLintHandler.report(_l -> warnDeprecated(pos.get(), s));
+            deferredLintHandler.report(lint2 -> warnDeprecated(lint2, pos.get(), s));
         }
     }
 
@@ -3850,7 +3850,7 @@ void checkPreview(DiagnosticPosition pos, Symbol other, Type site, Symbol s) {
 
     void checkRestricted(DiagnosticPosition pos, Symbol s) {
         if (s.kind == MTH && (s.flags() & RESTRICTED) != 0) {
-            deferredLintHandler.report(_l -> warnRestrictedAPI(pos, s));
+            deferredLintHandler.report(lint2 -> warnRestrictedAPI(lint2, pos, s));
         }
     }
 
@@ -4122,7 +4122,7 @@ void checkDivZero(final DiagnosticPosition pos, Symbol operator, Type operand) {
             int opc = ((OperatorSymbol)operator).opcode;
             if (opc == ByteCodes.idiv || opc == ByteCodes.imod
                 || opc == ByteCodes.ldiv || opc == ByteCodes.lmod) {
-                deferredLintHandler.report(_ -> lint.logIfEnabled(log, pos, LintWarnings.DivZero));
+                deferredLintHandler.report(lint2 -> lint2.logIfEnabled(log, pos, LintWarnings.DivZero));
             }
         }
     }
@@ -4135,8 +4135,8 @@ void checkDivZero(final DiagnosticPosition pos, Symbol operator, Type operand) {
      */
     void checkLossOfPrecision(final DiagnosticPosition pos, Type found, Type req) {
         if (found.isNumeric() && req.isNumeric() && !types.isAssignable(found, req)) {
-            deferredLintHandler.report(_ ->
-                lint.logIfEnabled(log, pos, LintWarnings.PossibleLossOfPrecision(found, req)));
+            deferredLintHandler.report(lint2 ->
+                lint2.logIfEnabled(log, pos, LintWarnings.PossibleLossOfPrecision(found, req)));
         }
     }
 
@@ -4335,8 +4335,8 @@ void checkDefaultConstructor(ClassSymbol c, DiagnosticPosition pos) {
                             // Warning may be suppressed by
                             // annotations; check again for being
                             // enabled in the deferred context.
-                            deferredLintHandler.report(_ ->
-                                lint.logIfEnabled(log, pos, LintWarnings.MissingExplicitCtor(c, pkg, modle)));
+                            deferredLintHandler.report(lint2 ->
+                                lint2.logIfEnabled(log, pos, LintWarnings.MissingExplicitCtor(c, pkg, modle)));
                         } else {
                             return;
                         }
@@ -4670,26 +4670,26 @@ private void checkVisible(DiagnosticPosition pos, Symbol what, PackageSymbol inP
 
     void checkModuleExists(final DiagnosticPosition pos, ModuleSymbol msym) {
         if (msym.kind != MDL) {
-            deferredLintHandler.report(_ ->
-                lint.logIfEnabled(log, pos, LintWarnings.ModuleNotFound(msym)));
+            deferredLintHandler.report(lint2 ->
+                lint2.logIfEnabled(log, pos, LintWarnings.ModuleNotFound(msym)));
         }
     }
 
     void checkPackageExistsForOpens(final DiagnosticPosition pos, PackageSymbol packge) {
         if (packge.members().isEmpty() &&
             ((packge.flags() & Flags.HAS_RESOURCE) == 0)) {
-            deferredLintHandler.report(_ ->
-                lint.logIfEnabled(log, pos, LintWarnings.PackageEmptyOrNotFound(packge)));
+            deferredLintHandler.report(lint2 ->
+                lint2.logIfEnabled(log, pos, LintWarnings.PackageEmptyOrNotFound(packge)));
         }
     }
 
     void checkModuleRequires(final DiagnosticPosition pos, final RequiresDirective rd) {
         if ((rd.module.flags() & Flags.AUTOMATIC_MODULE) != 0) {
-            deferredLintHandler.report(_ -> {
-                if (rd.isTransitive() && lint.isEnabled(LintCategory.REQUIRES_TRANSITIVE_AUTOMATIC)) {
+            deferredLintHandler.report(lint2 -> {
+                if (rd.isTransitive() && lint2.isEnabled(LintCategory.REQUIRES_TRANSITIVE_AUTOMATIC)) {
                     log.warning(pos, LintWarnings.RequiresTransitiveAutomatic);
                 } else {
-                    lint.logIfEnabled(log, pos, LintWarnings.RequiresAutomatic);
+                    lint2.logIfEnabled(log, pos, LintWarnings.RequiresAutomatic);
                 }
             });
         }
```
Turns out it's not quite that simple; the above patch breaks the build:
```
Building target 'test' in configuration 'macosx-aarch64-server-release'
Compiling up to 360 files for BUILD_jdk.compiler.interim
Updating support/src.zip
Compiling up to 368 files for jdk.compiler
/Users/archie/proj/jdk/src/jdk.compiler/share/classes/com/sun/tools/javac/file/FSInfo.java:142: warning: [deprecation] URL(URL,String) in URL has been deprecated
        URL retVal = new URL(base, input);
                     ^
error: warnings found and -Werror specified
/Users/archie/proj/jdk/src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java:269: warning: [restricted] Controller.enableNativeAccess(Module) is a restricted method.
            controller.enableNativeAccess(module);
                      ^
  (Restricted methods are unsafe and, if used incorrectly, might crash the Java runtime or corrupt memory)
1 error
2 warnings
make[3]: *** [/Users/archie/proj/jdk/build/macosx-aarch64-server-release/jdk/modules/jdk.compiler/_the.jdk.compiler_batch] Error 1
make[2]: *** [jdk.compiler-java] Error 2
```
So I don't know what's going on there exactly and have so far avoided that particular dark corner.

> I wonder if `enterDecl`/`leaveDecl` would be more evocative than push/pop. (but don't have any suggestion for the weird immediate mode).

Yes that naming would be better - if there were no immediate mode. But since immediate mode exists, the stack contains more than just declarations. Also, `push()` hopefully makes it more obvious that you are responsible putting in place a corresponding `pop()`.

> My general feeling here is that we'd be better off with a separate compilation step that runs after all the various front-end steps (e.g. after Flow).

I have that same feeling :) We have the new `warn()` compiler phase, so that would be the logical place to put the grand `flush()` operation.

I think this is do-able. Semantically things would get simpler, which means while the patch would be large, it would consist mostly of decomplexification. Rough sketch (this would include #23237):

* Move the deferral logic of `DeferredLintManager` into `Lint`; the `DeferredLintManager` singleton goes away.
* Lint has a single `flush()` method which is invoked _once_ per source file during the compiler `warn()` step
  * First, lexical deferrals are remapped to innermost containing `JCTree` 
  * Next, all warnings are emitted (in lexical order) by scanning the `JCCompilationUnit` and flushing each declaration
* Lint warnings are created via `Lint.logIfEnabled()` (or similar) at any time during compilation
  * If you can locate the warning with a `JCTree`, you supply that
  * Otherwise, you locate the warning with an integer `pos` source file offset
  * In either case, an assertion verifies that `flush()` has not yet been invoked

What do you think?

One thing I'm unclear of is whether `flush()` can assume that all of the accumulated warnings correspond to the same file. Might not be true with `CompilePolicy.BY_TODO`?

mcimadamore commented:
> I think this is do-able. Semantically things would get simpler, which means while the patch would be large, it would consist mostly of decomplexification. Rough sketch (this would come after #23237):
> 
>     * Move the deferral logic of `DeferredLintManager` into `Lint`; the `DeferredLintManager` singleton goes away.
> 
>     * Lint has a single `flush()` method which is invoked _once_ per source file during the compiler `warn()` step
>       
>       * First, lexical deferrals are remapped to innermost containing `JCTree`
>       * Next, all warnings are emitted (in lexical order) by scanning the `JCCompilationUnit` and flushing each declaration
> 
>     * Lint warnings are created via `Lint.logIfEnabled()` (or similar) at any time during compilation
>       
>       * If you can locate the warning with a `JCTree`, you supply that
>       * Otherwise, you locate the warning with an integer `pos` source file offset
>       * In either case, an assertion verifies that `flush()` has not yet been invoked
> 
> 
> What do you think?

This seems like a great plan. We don't have to get there in one step (and this PR can proceed separately). I just want to make sure that we end up somewhere around there :-)

(as your quick experiment with Check suggests, some of this stuff is not trivial at all -- even the use of the immediate mode seems a bit odd to be honest).

> 
> One thing I'm unclear of is whether `flush()` can assume that all of the accumulated warnings correspond to the same file. Might not be true with `CompilePolicy.BY_TODO`?

I believe that in `flow` we still see all the classes in the same source as belonging to the same toplevel unit. It's after desugar that things start to get messier. So, I believe that the warning visitor can assume that it's called once per compilation unit, and call flush on all the stuff that belongs to that unit.

archiecobbs commented:
> This seems like a great plan. We don't have to get there in one step (and this PR can proceed separately). I just want to make sure that we end up somewhere around there :-)

Sounds good. I created [JDK-8348611](https://bugs.openjdk.org/browse/JDK-8348611) to capture the idea.

> I believe that in `flow` we still see all the classes in the same source as belonging to the same toplevel unit. It's after desugar that things start to get messier. So, I believe that the warning visitor can assume that it's called once per compilation unit, and call flush on all the stuff that belongs to that unit.

Thanks - that makes life simpler.

mcimadamore reviewed (commented):
Question: what happens if we get rid of the "immediate" and we always defer? Does this only affect the order in which diagnostics are reported, or is there something deeper?

archiecobbs commented:
> Question: what happens if we get rid of the "immediate" and we always defer? Does this only affect the order in which diagnostics are reported, or is there something deeper?

I learn a little bit more about this. It looks like the only time immediate mode is used explicitly (i.e., other due to `DeferredLintHandler`'s default state) is when resolving imports ([here](https://github.com/openjdk/jdk/blob/89e5e7ab73472b7d02aac5b8b0c7e9f26db6ec32/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java#L368)). In that case, of course, there can be no enclosing `@SuppressWarnings` so that makes sense.

It looks like the other apparent use ([here](https://github.com/openjdk/jdk/blob/89e5e7ab73472b7d02aac5b8b0c7e9f26db6ec32/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Annotate.java#L257)) never occurs because `annotateLater()` is never invoked with a null `deferPos` (I've cleaned this up and added an assertion).

> This seems like a great plan. We don't have to get there in one step (and this PR can proceed separately). I just want to make sure that we end up somewhere around there :-)

I did some more exploration of this idea. The main trip-up at this point is the speculative compilation stuff, because in that case warnings that occur as a side effect should be discarded because they're not "real". This is done via `Log.DeferredDiagnosticHandler`. So my initial strategy of "Just defer all warnings and actually execute them later" is too simplistic (although I do have a prototype which "almost works" I'm happy to share). If/when you have time I'd like to discuss how to proceed with you. That discussion probably belongs somewhere else e.g. on compiler-dev or a draft PR.

mcimadamore commented:
> learn a little bit more about this. It looks like the only time immediate mode is used explicitly (i.e., other due to `DeferredLintHandler`'s default state) is when resolving imports ([here](https://github.com/openjdk/jdk/blob/89e5e7ab73472b7d02aac5b8b0c7e9f26db6ec32/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java#L368)). In that case, of course, there can be no enclosing `@SuppressWarnings` so that makes sense.

Can we use a similar "flush" strategy ? E.g. make the warnings be part of an import declaration, and then process them as part of a flush?

mcimadamore reviewed (approved):
Regardless of the separate discussion re. "pushImmediate", this PR contains a lot of good changes, and I'm ok with it being integrated as is. We can see if we can get rid of the "immediate" state in a separate PR.

mcimadamore commented:
> > learn a little bit more about this. It looks like the only time immediate mode is used explicitly (i.e., other due to `DeferredLintHandler`'s default state) is when resolving imports ([here](https://github.com/openjdk/jdk/blob/89e5e7ab73472b7d02aac5b8b0c7e9f26db6ec32/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java#L368)). In that case, of course, there can be no enclosing `@SuppressWarnings` so that makes sense.
> 
> Can we use a similar "flush" strategy ? E.g. make the warnings be part of an import declaration, and then process them as part of a flush? While it's good that you knocked one of the two usages off -- it's a bit sad to have this machinery "just" for one case.
> 
> Separately, I wonder if warnings issued from import statements should be suppressible as well (perhaps piggy backing on the toplevel class `@SuppressWarnings` uhmmm). It seems bad that there's some warnings that can't be suppressed programmatically.

I think it would be useful to understand which warnings can be generated during `resolveImports`. @jddarcy  pointed me at:

https://openjdk.org/jeps/211

See related change:

https://hg.openjdk.org/jdk9/jdk9/langtools/rev/82384454947c

Where we explicitly disable deprecation warnings on import since 9. Is there any other?

archiecobbs commented:
Hi @mcimadamore,

Thanks for the review & comments!

> Can we use a similar "flush" strategy ? E.g. make the warnings be part of an import declaration, and then process them as part of a flush? While it's good that you knocked one of the two usages off -- it's a bit sad to have this machinery "just" for one case.

Agreed - it seems to be a result of mixing the ordering of attributing things and flushing the `DeferredLintHandler`. An example is this program:
```java
import java.security.PolicySpi;
public class A {

    @SuppressWarnings("deprecation")
    private PolicySpi foo;

    @SuppressWarnings("deprecation")
    public void m() throws Exception {
        A.class.newInstance();
    }
}
```
If you comment out the one remaining `deferredLintHandler.pushImmediate()` invocation (in `TypeEnter.java`), then watch watch happens when compiling the above program, you'll see that `checkDeprecated()` is invoked twice for `PolicySpi`, and the first time (`other` = `unnamed package`) the current `Lint` instance is correct but the deferred instance is incorrect, and the second time (`other` = `A`) the reverse is true; when it's invoked for `newInstance()` (only once), the current instance is correct but the deferred instance is incorrect. So - all still very confusing to me.

"When to flush" is indeed a key issue for other reasons too. For example, it appears that `JCTree` nodes tend to randomly disappear after attribution (well, it only seems random to me so far), so that deferring warning analysis for too long can set you up for a `NullPointerException` chasing exercise. A simple example is in `Gen.genClass()` where all of a class' declarations are discarded after the classfile is generate (whether that matters depends on the compiler execution mode), and there are others which I'm still chasing down.

> I think it would be useful to understand which warnings can be generated during resolveImports. @jddarcy pointed me at: https://openjdk.org/jeps/211

I ran across this as well and accommodated it my prototype by effectively pretending there is a `@SuppressWarnings("deprecation", "removal", "preview")` on every `import` statement if source ≥ 9. This is possible because the `@SuppressWarning` analysis is a completely separate step.

> Separately, I wonder if warnings issued from import statements should be suppressible as well (perhaps piggy backing on the toplevel class `@SuppressWarnings` uhmmm). It seems bad that there's some warnings that can't be suppressed programmatically.

It would be nice if the compiler allowed `@SuppressWarnings` in more places. If/when it ever does, the proposed refactoring would hopefully make it easier to implement.

archiecobbs commented:
/integrate
